### PR TITLE
Group 6

### PR DESCRIPTION
**Pull request by koakekuna**
Group Members:
Kekoa Kekuna
Anshul Chauhan

### CODE REVIEW

File: calculator/main.py [hunk 1]
```diff
@@ -58,18 +72,26 @@ def get_history(calc):
...
-    pass
+    if not calc['history']:
+        return None
+    else:
+        operation = calc['history'][-1][1]
+        params = calc['history'][-1][2]
+        
+        # Argument Unpacking
+        # https://docs.python.org/3/tutorial/controlflow.html#unpacking-argument-lists
+        return calc['operations'][operation](*params)
```
Comment (jsymons) [inline calculator/main.py:97]: Since you're already storing the result of the operation in the history you don't have to re-run the calculation. You could just do:
`return calc['history'][-1][3]`

**Comment by jsymons** [comment]
All done. Looks good!

File: calculator/main.py [hunk 2]
```diff
@@ -26,8 +26,18 @@ def perform_operation(calc, operation, params):
...
-
+    # perform operation
+    for i in params:
+        if not isinstance(i, (int, float)):
+            raise InvalidParams("Given params are invalid.")
+    total = calc['operations'][operation](*params)
+    
+    # get the execution date-time and append in the history list
+    # double brackets since 'append' takes only one argument
+    calc['history'].append((str(datetime.now()), operation, params, total))
```
Comment (janiceshiu) [inline calculator/main.py:37]: huh. didn't realise that the default for `datetime.now()` is the format given in the test @jsymons we could've avoided some work but on the bright side we learnt about `strftime` and its formatting parameters

File: calculator/main.py [hunk 3]
```diff
@@ -58,18 +72,26 @@ def get_history(calc):
...
-    pass
+    return calc['history']
 
 
 def reset_history(calc):
     """
     Resets the calculator history back to an empty list.
     """
-    pass
+    calc['history'][:] = []
```
Comment (janiceshiu) [inline calculator/main.py:82]: could you explain why you used `[:]` please? `calc['history'] = []` is sufficient

File: calculator/main.py [hunk 4]
```diff
@@ -58,18 +72,26 @@ def get_history(calc):
...
 
 
 def repeat_last_operation(calc):
     """
     Returns the result of the last operation executed in the history.
     """
-    pass
+    if not calc['history']:
```
Comment (janiceshiu) [inline calculator/main.py:89]: ooh good point. we forgot about this edge case in ours (probably because it wasn't in the test suite, oops!)

File: calculator/main.py [hunk 5]
```diff
@@ -37,14 +47,18 @@ def add_new_operation(calc, operation):
     :param operation: Dict with the single operation to be added.
                       ie: {'add': add_function}
     """
-    pass
-
+    # validate operation is a dictionary
```
Comment (janiceshiu) [inline calculator/main.py:50]: Since this is a class, I like that you commented your code because it'll be good to look back at in the future.

File: calculator/exceptions.py [hunk 6]
```diff
@@ -4,4 +4,4 @@ class InvalidOperation(Exception):
 
 
 class InvalidParams(Exception):
-    pass
+    pass
```
Comment (janiceshiu) [inline calculator/exceptions.py:7]: imho normally there is a newline at the end of each file.

File: calculator/main.py [hunk 7]
```diff
@@ -26,8 +26,18 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
-
+    # perform operation
+    for i in params:
```
Comment (janiceshiu) [inline calculator/main.py:30]: might want to put something more descriptive than `i`. `for arg in params` or something, perhaps? so that future readers of your code (may include yourself) can see at a glance what each individual element is.

File: calculator/operations.py [hunk 8]
```diff
@@ -1,19 +1,20 @@
+from functools import reduce
```
Comment (janiceshiu) [inline calculator/operations.py:1]: nice. learning how to use and import other libraries

**Review by janiceshiu — CHANGES_REQUESTED** [review]
First - your formatting and naming could be a bit better (remember, code's written once and then read multiple times).

However, nice job overall. I especially liked that there were a lot of comments and that where implementation might be a bit unclear, you linked the relevant SO question.

File: calculator/main.py (hunk 7 above)
Comment (anshulxyz) [reply calculator/main.py:30]: Yeah, you're right. I keep forgetting that. :disappointed:

File: calculator/main.py (hunk 2 above)
Comment (y47bears) [reply calculator/main.py:37]: Our team wasn't aware of the default formatting for datetime.now(), either. However, knowing about strftime is VERY helpful. :)
